wordpress-mobile/WordPress-FluxC-Android
# Handle authentication for self-hosted sites in ReactNativeStore

Adding authentication handling for the rest API using a [nonce endpoint](https://developer.wordpress.org/reference/functions/wp_ajax_rest_nonce/). The flow looks like this:

1. If we have a rest endpoint saved use that, otherwise use discovery to find the rest endpoint;
2. Attempt to make the requested call;
3. If the call fails because the endpoint is "not found" and we are using a previously saved rest endpoint, clear the previously saved rest endpoint from storage and use discovery to find the rest endpoint and try the call again;
4. If the call fails because of a lack of authorization, clear any saved nonce and try to get a new nonce from the nonce endpoint. If the attempt to get a new nonce endpoint fails, use `wp-json/` as a fallback endpoint, and try again the call.

As part of this PR, I have simplified the external interface for these calls. Instead of having separate methods for WP.com and self-hosted sites, we now just have a single method that accepts a site and a path; the logic for appropriately handling that is now on the FluxC side (as a result, the [WPAndroid side](https://github.com/wordpress-mobile/WordPress-Android/pull/11522) is much simpler now).

### Remaining Issues to Address

#### Only try nonce endpoint if WP ≥ 5.3

It would be good if we only tried to use the nonce endpoint if the version of WordPress running on the site was ≥ 5.3 since that endpoint doesn't exist in WP < 5.3. I couldn't see an obvious way to check this however (for example, [`SiteModel.mSoftwareVersion`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/7e50af2dc11cd5587dc497826c1dd6994c18ced4/fluxc/src/main/java/org/wordpress/android/fluxc/model/SiteModel.java#L63) looked like it might be promising, but it was always `null` in my limited testing). Under the current implementation, after we fail to fetch a nonce for any reason other than a lack of an internet connection we won't try again in the next 5 minutes (or until the app is restarted since I'm just storing the nonce in memory).

Although not a part of this PR, in the future when WP < 5.3 (or when the nonce endpoint fails) we could try to scrape the nonce from the editor page. This is obviously a pretty big hack 😬 , but it has been tested out with some success [on iOS](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/202/files#diff-23bcb2622a43c4aa81dc5ab5c4540fccR204-R215).

#### Handling persistent failures to fetch nonce

~If our attempts to fetch a nonce for a particular site consistently fail, this implementation will currently try to retrieve the nonce again with every new request. This obviously is not a good user experience. iOS is currently addressing this by [only storing the nonce in memory for the life of the app](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/202/files#diff-23bcb2622a43c4aa81dc5ab5c4540fccR131), so only the first call after an app restart will try to fetch the nonce. This seems like a reasonable solution (especially since it is relatively simple), but it does require special handling if the attempt to retrieve the nonce fails because there is no internet connection (and maybe other scenarios I haven't thought of).~

~I haven't moved forward with adding any special retry handling yet because I would like to get some more thoughts on the best way to handle persistence and retrying.~ I'm moving forward with the same in-memory cache approach iOS has used because I'm noticing that sometimes get a 403 Forbidden response when trying to reuse the nonce after restarting the app.

### To Test

#### Insure new e2e tests pass

Run `./gradlew cAT -Pandroid.testInstrumentationRunnerArguments.class=org.wordpress.android.fluxc.release.ReleaseStack_ReactNativeWPAPIRequestTest`

#### Test in WPAndroid

1. Check out the related WPAndroid PR (https://github.com/wordpress-mobile/WordPress-Android/pull/11522)
2. Open a new post with the block editor
3. Add an image block
4. Add an image to that block from your WordPress Media Library
5. Open the settings for the image block
6. Insure that you can change the image size (thumbnail images should appear smaller and the block's HTML should reflect the updated image size and image URL). If the call fails, there will be no option to change the image size.
